internal/controller/promotion/composite.go

```diff
@@ -41,30 +43,87 @@ func (c *compositeMechanism) GetName() string {
 func (c *compositeMechanism) Promote(
 	ctx context.Context,
 	stage *kargoapi.Stage,
+	promo *kargoapi.Promotion,
 	newFreight kargoapi.SimpleFreight,
-) (kargoapi.SimpleFreight, error) {
+) (*kargoapi.PromotionStatus, kargoapi.SimpleFreight, error) {
 	if stage.Spec.PromotionMechanisms == nil {
-		return newFreight, nil
+		return &kargoapi.PromotionStatus{Phase: kargoapi.PromotionPhaseSucceeded},
+			newFreight, nil
 	}
 
+	var newStatus *kargoapi.PromotionStatus
 	newFreight = *newFreight.DeepCopy()
 
 	logger := logging.LoggerFromContext(ctx)
 	logger.Debugf("executing %s", c.name)
 
 	for _, childMechanism := range c.childMechanisms {
 		var err error
-		newFreight, err = childMechanism.Promote(ctx, stage, newFreight)
+		var otherStatus *kargoapi.PromotionStatus
+		otherStatus, newFreight, err = childMechanism.Promote(ctx, stage, promo, newFreight)
 		if err != nil {
-			return newFreight, errors.Wrapf(
+			return nil, newFreight, errors.Wrapf(
 				err,
 				"error executing %s",
 				childMechanism.GetName(),
 			)
 		}
+		newStatus = aggregateGitPromoStatus(newStatus, *otherStatus)
```

Comment:
I might be overlooking something here, so apologies in advance if I am...

I'm not sure we need this `aggregateGitPromoStatus` function. Since the only way to continue on to another iteration of the loop is for the current iteration to have succeeded, the severity of `otherStatus`'s phase is _always_ going to be >= that of `newStatus`, meaning we could skip any fancy logic for collapsing two statuses into one.

Or am I missing something?

Reply:
You are right, for this case where we short-circuit the composite mechanism as soon as we get a `Failed/Errored/Running` status it is unecessary. But `aggregateGitPromoStatus` is also called in `git::Promote`, where short-circuiting does not happen.

At the very least, `aggregateGitPromoStatus` in composite.go is unnecessary and should be removed. Good catch!

As for git.go, I think I should either short-circuit the loop (and remove`aggregateGitPromoStatus` altogether) or continue aaggregation without short-circuiting. Any preference?

Reply:
No preference other than whatever you feel ends up being simpler.

Reply:
Turns out I still need this because aggregateGitPromoStatus also has the responsibility of merging the metadata map.